clang/lib/Parse/ParseDecl.cpp

```diff
@@ -3708,9 +3709,9 @@ void Parser::ParseDeclarationSpecifiers(
             continue;
 
           if (PA.getKind() == ParsedAttr::AT_LifetimeBound)
-            Diag(PA.getLoc(), diag::err_attribute_wrong_decl_type_str)
```

Comment:
Do we have any instances of `_str` versions left?

- If not, maybe remove them from the TableGen files?
- If yes, what's stopping us from replacing them with enumerations and removing them from TableGen files?

Reply:
There are callsites left there: https://github.com/llvm/llvm-project/blob/dff7ef2353fec9f1006895c0e99bde704296eaa9/clang/utils/TableGen/ClangAttrEmitter.cpp#L4261 . These are less obvious to fix, and also the strings there are autogenerated themselves (based on other tablegens if I understand correctly).

Reply:
Makes sense, I didn't see this callsite. Thanks for the expalanation.

I guess one way would be to actually generate the enum declaration for subjects through TableGen too and also reuse the same declarations to point at the right diagnostic.
However, it's definitely a bigger refactoring and this change is already making things better, so let's land this!

Reply:
Yeah there's a FIXME in that code to somehow move this logic (the one that generates strings like `"tblgen-string-1" and "tblgen-string-2"`) to the diagnostics engine, however, as you said, this seems to be a significantly more complicated task: https://github.com/llvm/llvm-project/blob/dff7ef2353fec9f1006895c0e99bde704296eaa9/clang/utils/TableGen/ClangAttrEmitter.cpp#L4134

I won't overpromise that I'll be able to tackle this soon unless it keeps popping up. From your comment, it seems you're OK with this, but please let me know otherwise.

Reply:
Yeah, sorry for not being clear.
I am definitely not pushing for doing this refactoring now, it seems like too much work for relatively little benefit.
There are many things which aren't perfect, I fully agree we should leave this one be and move on to other things.